system/jlib/jstring.cpp

```diff
@@ -939,25 +939,73 @@ StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char
 {
     if (lenSource)
     {
-        size_t left = lenSource;
-        while (left >= lenOldStr)
+        if (lenOldStr)
         {
-            if (memcmp(source, oldStr, lenOldStr)==0)
+            // Scan for match on first character
+            size_t offset = 0;
+            size_t left = lenSource;
+            while (left >= lenOldStr)
             {
-                result.append(lenNewStr, newStr);
-                source += lenOldStr;
-                left -= lenOldStr;
+                if (source[offset] != oldStr[0])
+                {
+                    offset++;
+                    left--;
+                }
+                else
+                    break;
             }
-            else
+
+            // Reserve space when the new string is longer than the old string
+            unsigned steps = lenSource - lenOldStr - offset + 1;
+            unsigned maxResultLength = lenNewStr > lenOldStr ? lenSource + steps * (lenNewStr - lenOldStr) : lenSource;
+            result.ensureCapacity(maxResultLength);
+
+            // Add any characters before the match
+            result.append(offset, source);
+
+            size_t unmatchedChars = 0;
+            while (left >= lenOldStr)
             {
-                result.append(*source);
-                source++;
-                left--;
+                if (memcmp(source + offset, oldStr, lenOldStr)==0)
+                {
+                    if (unmatchedChars)
+                    {
+                        // Copy mismatched characters in blocks
+                        result.append(unmatchedChars, source + offset - unmatchedChars);
+                        unmatchedChars = 0;
+                    }
+                    result.append(lenNewStr, newStr);
+                    offset += lenOldStr;
+                    left -= lenOldStr;
+                }
+                else
+                {
+                    unmatchedChars++;
+                    offset++;
+                    left--;
+
+                    // Skip ahead to the next possible match
+                    while (left >= lenOldStr)
+                    {
+                        if (source[offset] != oldStr[0])
+                        {
+                            unmatchedChars++;
+                            offset++;
+                            left--;
+                        }
+                        else
+                            break;
+                    }
+                }
             }
-        }
 
-        // there are no more possible replacements, make sure we keep the end of the original buffer
-        result.append(left, source);
+            // If there were any characters left or characters that didn't match, append them
+            left += unmatchedChars;
+            if (left && (offset - left) < lenSource)
+                result.append(left, source + offset - unmatchedChars);
```

Comment:
This may be clearer as:

result.append(left, source + lenSource - left);

Assuming that what I typed here without checking actually is equivalent; please double-check.

Reply:
Changed to result.append(left, source + lenSource - left) as lenSource - left = offset - unmatchedChars.